src/lib/util/options.cpp

```diff
@@ -64,244 +64,484 @@ namespace
 	}
 };
 
+//**************************************************************************
+//  OPTIONS EXCEPTION CLASS
+//**************************************************************************
+
+//-------------------------------------------------
+//  options_exception - constructor
+//-------------------------------------------------
+
+options_exception::options_exception(std::string &&message)
+	: m_message(std::move(message))
+{
+}
+
+
+//-------------------------------------------------
+//  options_warning_exception - constructor
+//-------------------------------------------------
+
+options_warning_exception::options_warning_exception(std::string &&message)
+	: options_exception(std::move(message))
+{
+}
+
+
+//-------------------------------------------------
+//  options_error_exception - constructor
+//-------------------------------------------------
+
+options_error_exception::options_error_exception(std::string &&message)
+	: options_exception(std::move(message))
+{
+}
+
 
 //**************************************************************************
-//  CORE OPTIONS ENTRY
+//  CORE OPTIONS ENTRY BASE CLASS
 //**************************************************************************
 
 //-------------------------------------------------
 //  entry - constructor
 //-------------------------------------------------
 
-core_options::entry::entry(const char *name, const char *description, uint32_t flags, const char *defvalue)
-	: m_next(nullptr),
-		m_flags(flags),
-		m_error_reported(false),
-		m_priority(OPTION_PRIORITY_DEFAULT),
-		m_description(description),
-		m_changed(false)
+core_options::entry::entry(std::vector<std::string> &&names, core_options::option_type type, const char *description)
+	: m_names(std::move(names))
+	, m_priority(OPTION_PRIORITY_DEFAULT)
+	, m_type(type)
+	, m_description(description)
 {
-	// copy in the name(s) as appropriate
-	if (name != nullptr)
+	if (type == option_type::HEADER)
+		assert(m_names.size() == 0);
+	else
+		assert(m_names.size() > 0);
```

Comment:
This could be expressed as `assert(m_names.empty() == (m_type == option_type::HEADER))`